### ISSUE TITLE
Create a function type INVOKE () of unity in godot engine c #.

### ISSUE THREAD

**Issue by kone9** [godotengine/godot-proposals#44]
**Describe the project you are working on:**
I am currently not working on anything special, I am testing GODOT c # and making comparisons with UNITY 3D. I am mainly checking the corroutines and making a comparison with the INVOKE function of unity3D.

SPANISH:
Actualmente no estoy trabajando en nada especial,estoy probando GODOT c# y haciendo comparaciones con UNITY 3D.Principalmente estoy verificando las corrutinas y haciendo una comparación con la funcion INVOKE de unity3D.

**Describe how this feature / enhancement will help your project:**
Create a function type INVOKE () of unity in godot engine c #.
Invoke () is a function that allows us to call another function from time to time, it is very useful and although in Godot c # we have the corroutines, it would be nice to have a simple method like this to make things a bit easier. I leave a 3 minute video with a practical example of the procedure in unity.
https://www.youtube.com/watch?time_continue=160&v=sLYcVjWG2O8

SPANISH:
Crear un funcion tipo INVOKE() de unity en godot engine c#.
Invoke() es una función que nos permite llamar a otra función cada cierto tiempo,es muy util y aunque en Godot c# tenemos las corrutinas,estaria bueno tener un metodo simple como este para fácilitar un poco las cosas.Dejo un video de 3 minutos con un ejemplo practico del procedimiento en unity.
https://www.youtube.com/watch?time_continue=160&v=sLYcVjWG2O8

**Show a mock up screenshots/video or a flow diagram explaining how your proposal will work:**
<img width="538" alt="Screenshot_2" src="https://user-images.githubusercontent.com/12847027/64475149-241fcd00-d155-11e9-97a8-a98862c0abf6.png">

**Describe implementation detail for your proposal (in code), if possible:**
<img width="462" alt="Screenshot_3" src="https://user-images.githubusercontent.com/12847027/64475269-f0de3d80-d156-11e9-98c3-02edc3b04979.png">

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**
It could be solved with a corroutine, but the idea is to have a function always for something everyday like calling a function every so often.

**Is there a reason why this should be core and not an add-on in the asset library?:**
When we make video games one of the most common things is the instance of objects from time to time. They can be bullets, enemies or whatever and doing it with corroutines is a bit more difficult to understand. That is why I think a function like this is needed inside of Godot Engine.

**SPANISH:**
Cuando hacemos videojuegos una de las cosas más comunes es la instancia de objetos cada cierto tiempo.Pueden ser balas,enemigos o lo que sea y hacerlo con corrutinas es un poco más difícil de entender.Por eso creo que es necesario una función como esta dentro de Godot Engine.

**Comment by giulianob** [comment]
You can already do this several ways like with a Timer node or a C# timer. You can call functions with `Call` if you're going from gdscript to/from C# so I'm not sure what value this adds.

**Comment by kone9** [comment]
> You can already do this several ways like with a Timer node or a C# timer. You can call functions with `Call` if you're going from gdscript to/from C# so I'm not sure what value this adds.

It is much more difficult with the timer you have to use conditions, signals and it is quite complicated, especially with C # .. In the case of invoke you call the function after a certain time and ready, without signals, without conditions and without so much pain head.XD

**Comment by kone9** [comment]
> You can already do this several ways like with a Timer node or a C# timer. You can call functions with `Call` if you're going from gdscript to/from C# so I'm not sure what value this adds.

Well, the timer works from time to time, however if you press the key many times fast it does not work. The timer does not run until you have stopped pressing and then it runs only once.
This is possibly a logic error that I cannot understand, however I would like the INVOKE function in godot with c # .. these are all the tests I am going to do with this ... greetings

Bueno el timer funciona cada cierto tiempo,sin embargo si presional la tecla muchas veces rapido no funciona.EL timer no se ejecuta hasta que no dejaste de presionar y luego se ejecuta 1 sola ves.
Esto posiblemente sea un error de logica que yo no puedo entender,sin embargo me gustaria la funcion INVOKE en godot con c#..estas son toda las pruebas que voy a hacer con esto..saludos


<img width="454" alt="Screenshot_4" src="https://user-images.githubusercontent.com/12847027/64478770-a7f0ae00-d183-11e9-9619-a9a432c8e865.png">

**Comment by nhydock** [comment]
your code has a logic error in this situation.  The problem you're trying to solve here would be better to move the Timer and timeout onto your bullet instance instead of keeping it in the factory, since the Timer is related to the lifetime state of your object specifically.  Having an Invoke function would do the same, but with the downside of the timer associated with the Invoke just being thrown into memory and losing all context of its purpose, making things harder to debug and follow.

I feel this kind of feature has the potential to degrade a game's code quality, as there are sensible and safer equivalents that already exist.

**Comment by Jummit** [comment]
This is how it's done in Godot:
```
spawn_coin(position)
yield(get_tree().create_timer(COIN_SPAWN_DELAY), "timeout")
```
A bit clunky, but it works.

**Comment by regakakobigman** [comment]
The way it's described in the proposal, we would need first class functions. These can be simulated with FuncRefs for now, or we can use other solutions with call(), yield(), timers, and signals.. I think these are just as easy to work with as an `invoke()` function would be :)

However, if you wanted to make your own `invoke()` function, maybe you could emulate it with funcrefs and timers? You can make it as complex as you want, like making it return a reference to the function and the timer for easy access, or just making it fire-and-forget.

You can simulate it in other ways with yield() and timers/signals like others have suggested, which are good solutions. However, you may be able to make your own `invoke()` function by using FuncRefs.

**Comment by neikeq** [comment]
We just need a helper timer awaiter so you can do something like this:

```
while (true) {
    Foo();
    await ForSeconds(5);
}
```

and then if you really want a `Invoke` function, you could create one yourself:

```
async void Invoke(Action action, int seconds) {
    while (true) {
        action();
        await ForSeconds(seconds);
    }
}

Invoke(Foo, 5);
```

**Comment by giulianob** [comment]
@neikeq If you're using async then you can do
```
while (true) {
  action();
  await Task.Delay(1000);
}
```

**Comment by neikeq** [comment]
That would work if you don't care it's not resumed during idle or physics frame.

**Comment by noidexe** [comment]
According to the docs https://docs.godotengine.org/en/stable/getting_started/scripting/c_sharp/c_sharp_differences.html you can do it in the same way you'd do it in gdscript

``` C#
if ( terremoto == true ) {
  await ToSignal(GetTree().CreateTimer(5), "timeout");
  DetenerTerremoto();
}
```

**Comment by Calinou** [comment]
Closing per @noidexe's comment. Note that this will not work on HTML5 yet until threads support is implemented: https://github.com/godotengine/godot/issues/34506
